be/src/exec/chunks_sorter_full_sort.cpp

```diff
@@ -73,23 +116,146 @@ Status ChunksSorterFullSort::_partial_sort(RuntimeState* state, bool done) {
 
         _sorted_chunks.emplace_back(std::move(sorted_chunk));
         _total_rows += _unsorted_chunk->num_rows();
-        _unsorted_chunk.reset();
+        _unsorted_chunk->reset();
+        _staging_unsorted_rows = 0;
+        _staging_unsorted_bytes = 0;
+        _staging_unsorted_chunks.clear();
     }
 
     return Status::OK();
 }
 
 Status ChunksSorterFullSort::_merge_sorted(RuntimeState* state) {
     SCOPED_TIMER(_merge_timer);
-
-    RETURN_IF_ERROR(merge_sorted_chunks(_sort_desc, _sort_exprs, _sorted_chunks, &_merged_runs));
+    _profiler->num_sorted_runs->set((int64_t)_sorted_chunks.size());
+    // In cascading merging phase, the height of merging tree is ceiling(log2(num_sorted_runs)) + 1,
+    // so when num_sorted_runs is 1 or 2, the height merging tree is less than 2, the sorted runs just be processed
+    // in at most one pass. there is no need to enable lazy materialization which eliminates non-order-by output
+    // columns's permutation in multiple passes.
+    if (_eager_materialized_slots.empty() || _sorted_chunks.size() < 3) {
+        _runtime_profile->add_info_string("LazyMaterialization", "false");
+        RETURN_IF_ERROR(merge_sorted_chunks(_sort_desc, _sort_exprs, _sorted_chunks, &_merged_runs));
+    } else {
+        _runtime_profile->add_info_string("LazyMaterialization", "true");
+        _split_lazy_and_eager_chunks();
+        _assign_ordinals();
+        RETURN_IF_ERROR(merge_sorted_chunks(_sort_desc, _sort_exprs, _eager_materialized_chunks, &_merged_runs));
+    }
 
     return Status::OK();
 }
 
+void ChunksSorterFullSort::_assign_ordinals() {
+    _chunk_idx_bits = (int)std::ceil(std::log2(_eager_materialized_chunks.size()));
```

Comment:
you can use `64 - __builtin_clzl(_eager_materialized_chunks.size() - 1)` if _eager_materialized_chunks is not empty

Reply:
A good suggestion, but here is not a performance bottleneck, ceil(log2(...)) is clear